x-pack/plugins/security/public/session/session_timeout.ts

```diff
@@ -0,0 +1,342 @@
+ * Session will be extended at most once this number of milliseconds while user activity is detected.
+ */
+export const EXTENSION_THROTTLE_MS = 60 * 1000;
+
+/**
+ * Route to get session info and extend session expiration
+ */
+export const SESSION_ROUTE = '/internal/security/session';
+
+export interface SessionState extends Pick<SessionInfo, 'expiresInMs' | 'canBeExtended'> {
+  lastExtensionTime: number;
+}
+
+export class SessionTimeout {
+  private channel?: BroadcastChannel<SessionState>;
+
+  private isVisible = document.visibilityState !== 'hidden';
+  private isExtending = false;
+
+  private sessionState$ = new BehaviorSubject<SessionState>({
+    lastExtensionTime: 0,
+    expiresInMs: null,
+    canBeExtended: false,
+  });
+  private subscription?: Subscription;
+
+  private warningToast?: Toast;
+
+  private stopActivityMonitor?: Function;
+  private stopVisibilityMonitor?: Function;
+  private removeHttpInterceptor?: Function;
+
+  private stopRefreshTimer?: Function;
+  private stopWarningTimer?: Function;
+  private stopLogoutTimer?: Function;
+
+  constructor(
+    private notifications: NotificationsSetup,
+    private sessionExpired: ISessionExpired,
+    private http: HttpSetup,
+    private tenant: string
+  ) {}
+
+  public start() {
+    if (this.http.anonymousPaths.isAnonymous(window.location.pathname)) {
+      return;
+    }
+
+    this.subscription = this.sessionState$
+      .pipe(skip(1), throttleTime(1000), tap(this.toggleEventHandlers))
+      .subscribe(this.resetTimers);
+
+    return this.fetchSessionInfo();
+  }
+
+  public stop() {
+    const nextState = {
+      lastExtensionTime: 0,
+      expiresInMs: null,
+      canBeExtended: false,
+    };
+    this.toggleEventHandlers(nextState);
+    this.resetTimers(nextState);
+    this.subscription?.unsubscribe();
+  }
+
+  /**
+   * Event handler that receives session information from other browser tabs.
+   */
+  private handleChannelMessage = (message: SessionState) => {
+    this.sessionState$.next(message);
+  };
+
+  /**
+   * HTTP request interceptor which ensures that API calls extend the session only if tab is
+   * visible.
+   */
+  private handleHttpRequest = (fetchOptions: HttpFetchOptionsWithPath) => {
+    // Ignore requests to external URLs
+    if (fetchOptions.path.indexOf('://') !== -1) {
+      return;
+    }
+
+    if (fetchOptions.asSystemRequest === undefined) {
+      return { asSystemRequest: !this.isVisible };
+    }
+  };
+
+  /**
+   * HTTP response interceptor which allows us to prevent extending the session manually if it has
+   * already been extended as part of an API call.
+   */
+  private handleHttpResponse = (httpResponse: HttpResponse) => {
+    // Ignore session endpoint which is already handled by fetch callback
+    if (httpResponse.fetchOptions.path === SESSION_ROUTE) {
+      return;
+    }
+
+    // Extend session unless we're dealing with a system request
+    if (httpResponse.fetchOptions.asSystemRequest === false) {
```

Comment:
This can (and usually is) undefined, so we should probably change this from an `iff` to an `if`:
```suggestion
    if (!httpResponse.fetchOptions.asSystemRequest) {
```

Reply:
I only want to extend the session if this has been explicitly set to be a user request. 

There can be situations, especially when the page loads / service starts, where requests are sent before our request intercepter has been registered but the response is received when our response interceptor is already active in which case we don't know how to handle the request.

Reply:
Oh! I didn't put two and two together and I assumed the request interceptor and response interceptor were completely independent. I did not realize one is intended to impact the other. Some additional comments here to clarify that would be appreciated then 🙏